doc/API.md

```diff
@@ -47,21 +48,160 @@ async ({ id, object }) => {
 });
 ```
 
-## Function with guards
+## Extended function declaration
 
-Use guards optionally. You can apply guard for one or a few arguments. You can
-apply it multiple times for one argument for complex checking or for combination
-of multiple arguments.
+Declaration elements (fields):
+- `caption` - method display name
+- `description` - method description
+- `parameters` (optional) - parameters declarative schema
+- `validate` (optional) - parameters imperative validation function
+- `timeout` (optional) - execution timeout in milliseconds
+- `method` - sync or async lambda implementing method
+- `returns` (optional) - returning value declarative schema
+- `assert` (optional) - returning value imperative assertion function
+- `examples` - array of examples
+- `example` - single example (shorthand)
+- `errors` - collection of possible error
 
 ```js
-async ({ name, city, age }) => {
-  name = guard(name, { type: 'string', default: 'Marcus' });
-  city = guard(city, { type: 'string', default: 'Roma' });
-  age = guard(age, { type: 'number', required: true });
+({
+  parameters: {
+    a: { type: 'number' },
+    b: { type: 'number' },
+  },
 
-  const result = { name, city, born: age + 121 };
+  validate: ({ a, b }) => {
+    if (a % 3 === 0) throw new Error('Expected `a` to be multiple of 3');
+    if (b % 5 === 0) throw new Error('Expected `b` to be multiple of 5');
+  },
 
-  guard(result.born, { type: 'number', required: true });
-  return result;
-};
+  timeout: 1000,
+
+  method: async ({ a, b }) => {
+    const result = a + b;
+    return result;
+  },
+
+  returns: { type: 'number' },
+
+  assert: ({ a, b }, result) => {
+    if (result < a && result < b) {
+      throw new Error('Result expected to be greater than parameters');
+    }
+  },
```

Comment:
Not sure why would we want to have a special option for this, why not just check this at the end of `method`?

Reply:
Just to separate contract validation and business-logic